Documentation/proposals/202304-agent-deployment-pattern-sidecar.md

```diff
@@ -0,0 +1,151 @@
+
+Following what other operators, e.g. OtelCollector and Istio, are doing already, we'll use [Admission Webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) to mutate pods with the Prometheus Agent sidecar.
+
+The Prometheus-Operator project has implemented an admission webhook already. All that is left to do is add another route for pod mutation.
+
+In `pkg/admission/admission.go`, we'll do the following changes:
+
+*code edited for improved readability*
+
+```go
+func (a *Admission) Register(mux *http.ServeMux) {
+	// already existing functions
+	mux.HandleFunc(sidecarInjectorPath, a.serveSidecarInjector)
+}
+
+func (a *Admission) serveSidecarInjector(w http.ResponseWriter, r *http.Request) {
+	a.serveAdmission(w, r, a.injectSidecar)
+}
+
+func (a *Admission) injectSidecar(ar v1.AdmissionReview) *v1.AdmissionResponse {
+	a.incrementCounter(a.sidecarInjectionTriggeredCounter)
+
+	if ar.Request.Resource != podGVR {
+		err := fmt.Errorf("expected resource to be Pod, but received %v", ar.Request.Resource)
+		level.Warn(a.logger).Log("err", err)
+		return toAdmissionResponseFailure("Unexpected resource kind", "pod", []error{err})
+	}
+
+	pod := &corev1.Pod{}
+	if err := json.Unmarshal(ar.Request.Object.Raw, pod); err != nil {
+		level.Info(a.logger).Log("msg", errUnmarshalAdmission, "err", err)
+		return toAdmissionResponseFailure(errUnmarshalAdmission, "pod", []error{err})
+	}
+
+	reviewResponse := &v1.AdmissionResponse{Allowed: true}
+
+	agents, err := shoulPodBeInjected(pod)
+	if err != nil {
+		// We don't block pod creation if not able to get agents
+		// Otherwise we could break the whole cluster.
+		level.Error(a.logger).Log("msg", "Failed to discover if pod should be injected", "err", err)
+		return reviewResponse
+	}
+
+	patches, err := generatePatchesForSidecarInjection(pod.Raw, agents)
+	if err != nil {
+		// We don't block pod creation if not able to generate patches
+		// Otherwise we could break the whole cluster.
+		level.Error(a.logger).Log("msg", "cannot generate patch for sidecar injection", "err", err)
+		return reviewResponse
+	}
+
+	pt := v1.PatchTypeJSONPatch
+	reviewResponse.PatchType = &pt
+	reviewResponse.Patch = []byte(fmt.Sprintf("[%s]", strings.Join(patches, ",")))
+	return reviewResponse
+}
+
+// shoulPodBeInjected checks if a given pod is selected by  PrometheusAgents through
+// ServiceMonitor/PodMonitor/Probe selectors.
+// If it should be injected, it returns the agents responsible for the injection.
+func shoulPodBeInjected(pod corev1.Pod) ([]monitoringv1alpha1.PrometheusAgent, error) {}
+
+// generatePatchesForSidecarInjection generates the patches necessary to inject the agent sidecar into the container
+// The agent parameter is used to configure extra container flags, remote-write configs and alikes
+func generatePatchesForSidecarInjection(raw []bytes, agents []monitoringv1alpha1.PrometheusAgent)
+```
+
+### Caveats
```

Comment:
I'd also mention
* Increased permissions for the operator (it can now mutate pods)
* Opportunity to block pod creation when the admission webhook service isn't available

Reply:
Regarding the webhook blocking pod creation, I believe we can choose to Allow pods in case of unavailability

https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy